lib/rdoc/parser/ruby.rb

```diff
@@ -169,2204 +147,937 @@ def initialize(top_level, file_name, content, options, stats)
-        :public
-      else
-        raise RDoc::Error, "Invalid visibility: #{tk.name}"
-      end
+  def with_container(container, singleton: false)
+    old_container = @container
+    old_visibility = @visibility
+    old_singleton = @singleton
+    @visibility = :public
+    @container = container
+    @singleton = singleton
+    unless singleton
+      @module_nesting.push container
 
-    return vis_type, vis, singleton
+      # Need to update module parent chain to emulate Module.nesting.
+      # This mechanism is inaccurate and needs to be fixed.
+      container.parent = old_container
+    end
+    yield container
+  ensure
+    @container = old_container
+    @visibility = old_visibility
+    @singleton = old_singleton
+    @module_nesting.pop unless singleton
   end
 
-  ##
-  # Look for the first comment in a file that isn't a shebang line.
-
-  def collect_first_comment
-    skip_tkspace
-    comment = ''.dup
-    comment = RDoc::Encoding.change_encoding comment, @encoding if @encoding
-    first_line = true
-    first_comment_tk_kind = nil
-    line_no = nil
-
-    tk = get_tk
-
-    while tk && (:on_comment == tk[:kind] or :on_embdoc == tk[:kind])
-      comment_body = retrieve_comment_body(tk)
-      if first_line and comment_body =~ /\A#!/ then
-        skip_tkspace
-        tk = get_tk
-      elsif first_line and comment_body =~ /\A#\s*-\*-/ then
-        first_line = false
-        skip_tkspace
-        tk = get_tk
-      else
-        break if first_comment_tk_kind and not first_comment_tk_kind === tk[:kind]
-        first_comment_tk_kind = tk[:kind]
-
-        line_no = tk[:line_no] if first_line
-        first_line = false
-        comment << comment_body
-        tk = get_tk
+  # Records the location of this +container+ in the file for this parser and
+  # adds it to the list of classes and modules in the file.
 
-        if :on_nl === tk then
-          skip_tkspace_without_nl
-          tk = get_tk
-        end
-      end
+  def record_location container # :nodoc:
+    case container
+    when RDoc::ClassModule then
+      @top_level.add_to_classes_or_modules container
     end
 
-    unget_tk tk
-
-    new_comment comment, line_no
-  end
-
-  ##
-  # Consumes trailing whitespace from the token stream
-
-  def consume_trailing_spaces # :nodoc:
-    skip_tkspace_without_nl
+    container.record_location @top_level
   end
 
-  ##
-  # Creates a new attribute in +container+ with +name+.
-
-  def create_attr container, single, name, rw, comment # :nodoc:
-    att = RDoc::Attr.new get_tkread, name, rw, comment, single == SINGLE
-    record_location att
-
-    container.add_attribute att
-    @stats.add_attribute att
-
-    att
-  end
+  # Scans this Ruby file for Ruby constructs
 
-  ##
-  # Creates a module alias in +container+ at +rhs_name+ (or at the top-level
-  # for "::") with the name from +constant+.
+  def scan
+    @tokens = RDoc::Parser::RipperStateLex.parse(@content).sort_by { |t| [t.line_no, t.char_no] }
```

Comment:
If you wanted, you could use `Prism.parse_lex`, which will give you a parse result with both the AST and the tokens. Alternatively, you could use `Prism.lex` or `Prism.lex_compat` to get back a list of tokens in a separate pass.

Reply:
Thank you for letting me know.
Ideally, we should use it, but `RDoc::Parser::RipperStateLex` is doing some concatenation of tokens and RDoc's source code colorization seems to depends on it.

I want to make that part out of scope of this pull request.

Reply:
That makes sense!